pkg/storage/snapshot/restore.go

```diff
@@ -552,10 +554,111 @@ func (t *vmRestoreTarget) Reconcile() (bool, error) {
+	if !backendstorage.IsBackendStorageNeededForVMI(&snapshotVM.Spec.Template.Spec) {
+		return true, nil
+	}
+
+	// Retrieve only the backend volume
+	volumes, err := storageutils.GetVolumes(snapshotVM, t.controller.Client, storageutils.WithBackendVolume)
+	if err != nil {
+		if storageutils.IsErrNoBackendPVC(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	pvc, err := t.controller.getPVC(snapshotVM.Namespace, volumes[0].VolumeSource.PersistentVolumeClaim.ClaimName)
+	if err != nil || pvc == nil {
+		return false, err
+	}
+
+	// Step 1: Remove backend pvc label from original backend PVC
+	if pvc.Labels != nil {
+		// Only remove label when the VM name is the same since the backend logic filters by VM name + label
+		if t.vmRestore.Spec.Target.Name == snapshotVM.Name {
+			isRestorePVC := false
+			for _, vr := range t.vmRestore.Status.Restores {
+				if vr.PersistentVolumeClaimName == pvc.Name {
+					isRestorePVC = true
+				}
+			}
+			if !isRestorePVC {
+				// Generate patch to remove the labels
+				patchBytes, err := patch.New(
+					patch.WithTest("/metadata/labels", pvc.Labels),
+					patch.WithReplace("/metadata/labels", getFilteredLabels(pvc.Labels)),
+				).GeneratePayload()
+
+				if err != nil {
+					return false, err
+				}
+				if _, err := t.controller.Client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Patch(context.Background(), pvc.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{}); err != nil {
+					return false, err
+				}
+			}
+		}
+	}
+
+	// Step 2: Update restore PVC with backend labels
+	for _, vr := range t.vmRestore.Status.Restores {
+		if vr.VolumeName == storageutils.BackendPVCVolumeName(t.vmRestore.Spec.Target.Name) {
+			restorePVC, err := t.controller.getPVC(t.vmRestore.Namespace, vr.PersistentVolumeClaimName)
+			if err != nil {
+				return false, err
+			}
+
+			patchSet := patch.New()
+
+			if restorePVC.Labels == nil {
+				patchSet.AddOption(
+					patch.WithAdd("/metadata/labels", map[string]string{
+						backendstorage.PVCPrefix: t.vmRestore.Spec.Target.Name,
+					}),
+				)
+			} else {
+				updatedLabels := make(map[string]string, len(restorePVC.Labels))
+				for k, v := range restorePVC.Labels {
+					updatedLabels[k] = v
+				}
+				updatedLabels[backendstorage.PVCPrefix] = t.vmRestore.Spec.Target.Name
+
+				patchSet.AddOption(
+					patch.WithTest("/metadata/labels", restorePVC.Labels),
+					patch.WithReplace("/metadata/labels", updatedLabels),
+				)
+			}
+
+			patchBytes, err := patchSet.GeneratePayload()
+			if err != nil {
+				return false, err
+			}
+
+			if _, err := t.controller.Client.CoreV1().PersistentVolumeClaims(restorePVC.Namespace).Patch(context.Background(), restorePVC.Name, types.JSONPatchType, patchBytes, metav1.PatchOptions{}); err != nil {
+				return false, err
+			}
+
+			// Step 3: Remove original backend PVC after the restore PVC has been updated
```

Comment:
I am a little concerned about deleting anything before the operation is "complete" meaning the spec is updated. That is why we delete old datavolumes later in the flow. Is there any way we can do that here? Maybe in addition to deleting the label above we can add another label for cleanup later?

Reply:
That makes sense for volumes that appear in the spec (like dvs), but in this case no changes in the backend PVC will affect the VM spec. The operation will be fully completed once the virt launcher pod starts using this PVC instead of the previous one, but removing the label from the old one and adding it to the new one is probably the most we can do in the controller. I would argue that's enough, but let me know what you think.

Reply:
Okay pretty sure deleting the PVC here is what's causing the test to flake. I'll find a better way to delete it.